String builder

stephentetley opened the pull request:
Hi Magnus, here is the pull request for review.

magnus-madsen commented on main/src/flix/runtime/library/MonoStringBuilder.java:
```diff
@@ -0,0 +1,102 @@
+/*
```
We usually use the word "mono" to refer to the process of monomorphization whereby a generic class is instead with its actual type arguments. The StringBuilder class is actually not generic, but it does use alot of the overloaded methods, so the naming is slightly misleading.

How do you feel about coming up with some general naming scheme for when we have to write the wrappers?

We could simply call them all `JStringBuilder` or `StringBuilderWrapper` or similar?

magnus-madsen replied:
The question is not too important, since at the some point we will hopefully have a better import mechanism :)

magnus-madsen commented on main/src/flix/runtime/library/MonoStringBuilder.java:
```diff
@@ -0,0 +1,102 @@
+import java.math.BigInteger;
+
+/**
+ * Wrap Java's StringBuilder, provide a small, monomorphic API.
```
A wrapper class for StringBuilder that provides concrete, non-overloaded methods.

magnus-madsen commented on main/src/library/String.flix:
```diff
@@ -71,4 +71,10 @@ namespace String {
+    ///
+    /// Get the system line separator.
+    ///
+    pub def lineSeparator() : Str =
```
I wonder if it would make sense to put this into some other namespaces? I guess we can leave it here for now...

magnus-madsen commented on main/src/library/StringBuilder.flix:
```diff
@@ -0,0 +1,139 @@
+opaque type StringBuilder = ##flix.runtime.library.MonoStringBuilder
+
+namespace StringBuilder {
+
```
Remove extra newlines.

magnus-madsen commented on main/src/library/StringBuilder.flix:
```diff
@@ -0,0 +1,139 @@
+    ///
+    /// Append the Str `s` to the StringBuffer `sb`.
+    ///
+    pub def appendString(sb: StringBuilder, s: Str): Unit =
```
In the future, I am planning to rename `Str` to `String` so having two methods `appendStr` and `appendString` might be confusing. Maybe we could call one of  them `appendToBuffer` or the like? I am open to suggestions.

magnus-madsen commented on main/src/library/StringBuilder.flix:
```diff
@@ -0,0 +1,139 @@
+        let StringBuilder(msb) = sb;
+        native method flix.runtime.library.MonoStringBuilder.length(msb) as Int32
+
+
```
Remove extra new lines.

magnus-madsen commented on main/test/ca/uwaterloo/flix/library/TestStringBuilder.flix:
```diff
@@ -0,0 +1,346 @@
+    let sb1 = StringBuilder.new();
+    let sb2 = StringBuilder.new();
+    StringBuilder.append(sb1, sb2);
+    StringBuilder.toString(sb1) `assertEq!` ""
```
I *think* we are moving away from using `assert!` and `assertEq`, instead a test case should simply return a true boolean if it was successful.

(I know this is not documented anywhere, and many tests still use it. The reasoning is that `assertEq` throws an exception, but we do not plan to have actual exceptions, so it is somewhat "magical" and I want to move away from that).

magnus-madsen commented on main/test/ca/uwaterloo/flix/library/TestStringBuilder.flix:
```diff
@@ -0,0 +1,346 @@
+// appendFloat32                                                           //
+/////////////////////////////////////////////////////////////////////////////
+@test
+def appendFloat3201(): Bool = 
```
What about NaN and Inf/-Inf?

(Probably these are also poorly supported in Flix already... so maybe a TODO will suffice. I guess you can get NaN from `0.0/0.0` and Inf in similar spirit.)

stephentetley commented:
Thanks Magnus, I'll look at making those changes today.

Apropos adding `lineSeparator()` to `String.flix` - there are a couple of functions that depend on it
in additions I intend to propose to `String` otherwise I would have made it a local function to `StringBuilder`.

I'll certainly add my name to authors, do you want me to change the copyright notice in the file headers to either to you or a collective name like "Flix Development team"? 

It could be confusing for subsequent developers wanting to contribute the project if individual files have different copyright authors.

magnus-madsen commented:
> Apropos adding lineSeparator() to String.flix - there are a couple of functions that depend on it
in additions I intend to propose to String otherwise I would have made it a local function to StringBuilder.

OK, I think that is reasonable then.

> I'll certainly add my name to authors, do you want me to change the copyright notice in the file headers to either to you or a collective name like "Flix Development team"?

You own the copyright, so I think its best to leave your name there. If others make changes, I guess they would hypothetically add their own name.

stephentetley commented:
Hi Magnus

I've pushed changes to the `string-builder` branch, is this effectively "updating" the pull request? This is the first proper pull request I've done.

Details: 

I think I've removed all double blank lines.

AUTHORS.md updated.

`MonoStringBuilder.java` is now `StringBuilderWrapped.java` and the headline comment changed.

StringBuilder API - `appendStr` / `appendString`. 
I've removed `appendStr` - if Flix's `Str` type is going to be renamed to `String` then `appendStr` would be superfluous.

The tests have been update to use `==` rather than `assertEq!`. 

I've added extra tests for NaN and the infinities to Float32 and Float64. The tests on Floats were intentionally a bit scant as I'm not sure what precision you should expect in the output, but as NaN and Infinity aren't show "numerically" the tests are simple. At some point StringBuilder ought to have something like `appendFloat{32/64}WithPrecision()` so the user can decide what to do.

magnus-madsen commented:
Looks great! Thanks for your work :)

magnus-madsen commented:
I have merged the PR using squash. You may have to rebase your fork to avoid conflicts.

stephentetley commented:
Thanks Magnus. I'll go back to working up the String and Array additions.

magnus-madsen commented:
Side note: I wonder if we could/should implement our own string builder using a simple mutable array of strings? That should already be supported.

stephentetley commented:
Hi Magnus

That's something to think about. With F# I've always had a String Builder available as part of the standard library and I haven't even thought about its implementation.

A pure String Builder is valuable when monads get more prevalent in your code (obviously it should be more efficient than simply collecting a list of strings). You need the String Builder to have purity (persistence) so the monads can have choice / backtracking. 

I've put a sketch in my `flix-doodle` repository to remind me to look at this at some point in future.

https://github.com/stephentetley/flix-doodle/blob/master/join-string/JoinString.flix

Because it is persistent the append functions return a JoinString rather mutate the existing one (and return Unit).

magnus-madsen commented:
What I was wondering was if we could implement our own string builder as a string array and an index into it. Append would then bump the index and place the element. If we reach the end of the array, we would copy it into an array of twice the size. Essentially the same mechanism used to implement ArrayLists. The advantage of doing this would be to have more code in Flix and less code in Java glue code.
